### PR TITLE
dev-util/ccache: prefer using bundled C++ deps by default

### PR DESCRIPTION
**Pull request by hhoffstaette**
This prevents potential problems with libstdc++ especially during toolchain builds.

Closes: https://bugs.gentoo.org/938011

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by hhoffstaette** [comment]
Also need to figure out how to handle CXX="ccache g++" situations.

**Comment by hhoffstaette** [comment]
> Also need to figure out how to handle CXX="ccache g++" situations.

Updated to sanitise CC/CXX as well.

**Comment by floppym** [comment]
I don't think it makes sense to disable ccache when building ccache. Any package could be affected by having a broken ccache installed, so I don't see why we should treat ccache itself any differently.

**Comment by hhoffstaette** [comment]
> I don't think it makes sense to disable ccache when building ccache. Any package could be affected by having a broken ccache installed, so I don't see why we should treat ccache itself any differently.

That's OK, the second commit was just the result of a thought experiment on IRC, since ignoring FEATURES turned out ot be not possible, and it went of from there in search of a strip-ccache toolchain function. The bug that prompted this would have been prevented by static linking (the first commit in this PR), so it's more a cherry on top. I believe we do something similar in distcc (which also came up in the discussion). In any case the first commit here is more important. If the second is too controversial for whatever reasons we can just skip it.
Also I don't know why one would enable ccache globally, but apparently people do, so..

**Comment by negril** [comment]
> I don't think it makes sense to disable ccache when building ccache. Any package could be affected by having a broken ccache installed, so I don't see why we should treat ccache itself any differently.

Because you can't rebuild `ccache` with broken `ccache`. So that just eases the recovery.

**Comment by mgorny** [comment]
You can't rebuild any of the dependencies with broken ccache, so that's hardly a meaningful reason. Adding hackarounds to one out of a dozen packages is just short-sighted, and to that, a hackaround that is actively harmful in normal situations (i.e. >99% cases).

**Comment by hhoffstaette** [comment]
I removed the second commit.

File: dev-util/ccache/ccache-4.10.2-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# Enable 'static-c++' by default to make 'gcc' ebuild Just Work: bug #761220
+IUSE="${MY_DOCS_USEFLAG} redis +static-c++ test"
+RESTRICT="!test? ( test )"
+
+# TODO: package NonstdSpan, TlExpected
+DEPEND="
+	>=app-arch/zstd-1.3.4:=
+	!static-c++? ( >=dev-cpp/cpp-httplib-0.10.6:= )
+	>=dev-libs/blake3-1.4.0:=
+	!static-c++? ( >=dev-libs/libfmt-8.0.0:= )
```
Comment (floppym) [inline dev-util/ccache/ccache-4.10.2-r1.ebuild:47]: This can be combined with the cpp-httplib dep into a single use-conditional block.

**Review by floppym — CHANGES_REQUESTED** [review]
Does `-static-libstdc++` work with clang (and libcxx)?

File: dev-util/ccache/ccache-4.10.2-r1.ebuild (hunk 1 above)
Comment (hhoffstaette) [reply dev-util/ccache/ccache-4.10.2-r1.ebuild]: done.

File: dev-util/ccache/ccache-4.10.2-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		-DENABLE_DOCUMENTATION=$(usex doc)
+		-DENABLE_TESTING=$(usex test)
+		-DDEPS=LOCAL
+		-DREDIS_STORAGE_BACKEND=$(usex redis)
+	)
+
+	use static-c++ && mycmakeargs+=(
+		-DDEP_CPPHTTPLIB=BUNDLED
+		-DDEP_FMT=BUNDLED
+	)
```
[the remark this answers is not in the document]
Comment (hhoffstaette) [reply dev-util/ccache/ccache-4.10.2-r1.ebuild]: done.

**Comment by hhoffstaette** [comment]
> Does `-static-libstdc++` work with clang (and libcxx)?

Yes, at least as far as clang goes. I added my clang.conf in package.env and it built fine, no dynamic C++ dependencies. I have not tested with libcxx though.

**Comment by ionenwks** [comment]
> > Does `-static-libstdc++` work with clang (and libcxx)?
> 
> Yes, at least as far as clang goes. I added my clang.conf in package.env and it built fine, no dynamic C++ dependencies. I have not tested with libcxx though.

Can confirm it's fine with libcxx too, just tried in my llvm-musl chroot that I use for testing edge cases problems.

And I'm actually surprised that this PR seem to have fixed an issue I was having, since clang/libcxx18 or so and USE=static-c++ (fine if I disabled it or used 17), I was getting:
```
$ ccache --version
libc++abi: terminating due to uncaught exception of type std::bad_cast: std::bad_cast
Aborted
```
However, with ccache-4.10.2-r1[static-c++]? Runs fine!

**Review by floppym — APPROVED** [review]
This seems reasonable to me.

**Pull request closed without merging**